README.md
# react-appointment-picker

> Component to pick appointments

[![NPM](https://img.shields.io/npm/v/react-appointment-picker.svg)](https://www.npmjs.com/package/react-appointment-picker) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier) [![Build Status](https://travis-ci.com/roggervalf/react-appointment-picker.svg?branch=master)](https://travis-ci.com/github/roggervalf/react-appointment-picker) [![NPM downloads](https://img.shields.io/npm/dm/react-appointment-picker)](https://www.npmjs.com/package/react-appointment-picker) [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Demo

[![Edit AppointmentPicker](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/eloquent-wave-qnxzv?fontsize=14)

## About

This react component is useful for schedules. Based in [react-seat-picker](https://www.npmjs.com/package/react-seat-picker).

## Install

```bash
npm install --save react-appointment-picker
```

Or

```bash
yarn add react-appointment-picker
```

## Usage

```jsx
import React, { Component } from 'react';

import { AppointmentPicker } from 'react-appointment-picker';

export default class App extends Component {
  state = {
    loading: false,
    continuousLoading: false
  };

  addAppointmentCallback = ({
    addedAppointment: { day, number, time, id },
    addCb
  }) => {
    this.setState(
      {
        loading: true
      },
      async () => {
        await new Promise((resolve) => setTimeout(resolve, 2000));
        console.log(
          `Added appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        addCb(day, number, time, id);
        this.setState({ loading: false });
      }
    );
  };

  removeAppointmentCallback = ({ day, number, time, id }, removeCb) => {
    this.setState(
      {
        loading: true
      },
      async () => {
        await new Promise((resolve) => setTimeout(resolve, 2000));
        console.log(
          `Removed appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        removeCb(day, number);
        this.setState({ loading: false });
      }
    );
  };

  addAppointmentCallbackContinuousCase = ({
    addedAppointment: { day, number, time, id },
    addCb,
    removedAppointment: params,
    removeCb
  }) => {
    this.setState(
      {
        continuousLoading: true
      },
      async () => {
        if (removeCb) {
          await new Promise((resolve) => setTimeout(resolve, 1000));
          console.log(
            `Removed appointment ${params.number}, day ${params.day}, time ${params.time}, id ${params.id}`
          );
          removeCb(params.day, params.number);
        }
        await new Promise((resolve) => setTimeout(resolve, 1000));
        console.log(
          `Added appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        addCb(day, number, time, id);
        this.setState({ continuousLoading: false });
      }
    );
  };

  removeAppointmentCallbackContinuousCase = (
    { day, number, time, id },
    removeCb
  ) => {
    this.setState(
      {
        continuousLoading: true
      },
      async () => {
        await new Promise((resolve) => setTimeout(resolve, 2000));
        console.log(
          `Removed appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        removeCb(day, number);
        this.setState({ continuousLoading: false });
      }
    );
  };

  render() {
    const days = [
      [
        { id: 1, number: 1, isSelected: true, periods: 2 },
        { id: 2, number: 2 },
        null,
        { id: 3, number: '3', isReserved: true },
        { id: 4, number: '4' },
        null,
        { id: 5, number: 5 },
        { id: 6, number: 6 }
      ],
      [
        { id: 7, number: 1, isReserved: true, periods: 3 },
        { id: 8, number: 2, isReserved: true },
        null,
        { id: 9, number: '3', isReserved: true },
        { id: 10, number: '4' },
        null,
        { id: 11, number: 5 },
        { id: 12, number: 6 }
      ],
      [
        { id: 13, number: 1 },
        { id: 14, number: 2 },
        null,
        { id: 15, number: 3, isReserved: true },
        { id: 16, number: '4' },
        null,
        { id: 17, number: 5 },
        { id: 18, number: 6 }
      ],
      [
        { id: 19, number: 1 },
        { id: 20, number: 2 },
        null,
        { id: 21, number: 3 },
        { id: 22, number: '4' },
        null,
        { id: 23, number: 5 },
        { id: 24, number: 6 }
      ],
      [
        { id: 25, number: 1, isReserved: true },
        { id: 26, number: 2 },
        null,
        { id: 27, number: '3', isReserved: true },
        { id: 28, number: '4' },
        null,
        { id: 29, number: 5 },
        { id: 30, number: 6, isReserved: true }
      ]
    ];
    const { loading, continuousLoading } = this.state;
    return (
      <div>
        <h1>Appointment Picker</h1>
        <AppointmentPicker
          addAppointmentCallback={this.addAppointmentCallback}
          removeAppointmentCallback={this.removeAppointmentCallback}
          initialDay={new Date('2018-05-05')}
          days={days}
          maxReservableAppointments={3}
          alpha
          visible
          selectedByDefault
          loading={loading}
        />
        <h1>Appointment Picker Continuous Case</h1>
        <AppointmentPicker
          addAppointmentCallback={this.addAppointmentCallbackContinuousCase}
          removeAppointmentCallback={
            this.removeAppointmentCallbackContinuousCase
          }
          initialDay={new Date('2018-05-05')}
          days={days}
          maxReservableAppointments={2}
          alpha
          visible
          selectedByDefault
          loading={continuousLoading}
          continuous
        />
      </div>
    );
  }
}
```

### Props

| Name                        | Type     | Default                                                                                                                                                                   | Required | Description                                                                                                                                                                                                                                |
| --------------------------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `alpha`                     | boolean  | `false`                                                                                                                                                                   | `false`  | Displays the name of the day of the week (`true`), otherwise in `dd-mm-yyyy` format.                                                                                                                                                       |
| `visible`                   | boolean  | `false`                                                                                                                                                                   | `false`  | Shows the day (`true`), otherwise they are hidden (`false`).                                                                                                                                                                               |
| `loading`                   | boolean  | `false`                                                                                                                                                                   | `false`  | Shows a white mask on the appointmentPicker.                                                                                                                                                                                               |
| `continuous`                | boolean  | `false`                                                                                                                                                                   | `false`  | Allows to continue select appointments while remove previous ones if you already have max reservable appointments.                                                                                                                         |
| `selectedByDefault`         | boolean  | `false`                                                                                                                                                                   | `false`  | Allow to have already selected appointments (`true`), otherwise (`false`) they aren't going to be checked by their isSelected property.                                                                                                    |
| `maxReservableAppointments` | number   | 0                                                                                                                                                                         | `false`  | Limits the number of selectable appointments.                                                                                                                                                                                              |
| `initialDay`                | Date     | -                                                                                                                                                                         | `true`   | Sets the initial day for your days.                                                                                                                                                                                                        |
| `unitTime`                  | number   | 15 _ 60 _ 1000                                                                                                                                                            | `false`  | Sets the minimal period of time between appointments.                                                                                                                                                                                      |
| `local`                     | string   | `en-US`                                                                                                                                                                   | `false`  | Sets the locale param for Dates variables. See [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString).                                                                  |
| `addAppointmentCallback`    | function | ({addedAppointment: {day, number, time, id}, addCb}) => { console.log(`Added appointment ${number}, day ${day}, time ${time}, id ${id}`); addCb(day, number, time, id);}, | `false`  | Should be customized as you need. Remember to use addCb(day,number,time,id) for accepting the selection, otherwise omit it. For continuous case see the example where should use removeCb(day,number) for previously selected appointment. |
| `removeAppointmentCallback` | function | ({day, number, time, id}, removeCb) => {console.log( `Removed appointment ${number}, day ${day}, time ${time}, id ${id}`); removeCb(day,number);}                         | `false`  | Should be customized as you need. Remember to use removeCb(day,number) for accepting the deselection, otherwise omit it.                                                                                                                   |
| `days`                      | array    | -                                                                                                                                                                         | `true`   | Array of arrays of json. (See next section).                                                                                                                                                                                               |

### Appointment properties

Each json in days prop could be `null` (empty appointment) or has these properties.

| Name         | Type             | Default   | Required | Description                                            |
| ------------ | ---------------- | --------- | -------- | ------------------------------------------------------ |
| `id`         | number or string | undefined | `false`  | It identify an appointment.                            |
| `number`     | number or string | undefined | `false`  | It will be its order.                                  |
| `isSelected` | boolean          | `false`   | `false`  | It will be checked in case selectedByDefault is true.  |
| `isReserved` | boolean          | `false`   | `false`  | Disable the option of click it.                        |
| `periods`    | number           | 1         | `false`  | Represents how many periods belongs to an appointment. |

## Contributing

Fork the repo, make some changes, submit a pull-request! Here is the [contributing](contributing.md) doc that has some details.

## License

MIT © [roggervalf](https://github.com/roggervalf)


CHANGELOG.md
## [3.2.1](https://github.com/roggervalf/react-appointment-picker/compare/v3.2.0...v3.2.1) (2022-05-24)


### Bug Fixes

* **deps:** update react peer deps for versions 17 and 18 ([#16](https://github.com/roggervalf/react-appointment-picker/issues/16)) ([2dbdb88](https://github.com/roggervalf/react-appointment-picker/commit/2dbdb8864ed2cfba38c5fdc91aa0cdc1d2ece6ce))

# [3.2.0](https://github.com/roggervalf/react-appointment-picker/compare/v3.1.0...v3.2.0) (2021-08-02)


### Features

* **appointment-picker:** transform to functional component ([2669a2f](https://github.com/roggervalf/react-appointment-picker/commit/2669a2f078886025827eba9b8d021c87a85fd2e4))

# [3.1.0](https://github.com/roggervalf/react-appointment-picker/compare/v3.0.1...v3.1.0) (2021-07-28)


### Features

* **components:** transform class components to functional components ([5f69d25](https://github.com/roggervalf/react-appointment-picker/commit/5f69d251c29fa4ba739023838dff81879236d03b))

## [3.0.1](https://github.com/roggervalf/react-appointment-picker/compare/v3.0.0...v3.0.1) (2021-01-09)


### Bug Fixes

* **codesandbox:** updating example project reference ([243f30b](https://github.com/roggervalf/react-appointment-picker/commit/243f30b0a078a2c4cb5592d3ad8b37385d46d63b)), closes [#12](https://github.com/roggervalf/react-appointment-picker/issues/12)

# [3.0.0](https://github.com/roggervalf/react-appointment-picker/compare/v2.0.1...v3.0.0) (2020-06-19)


### Features

* **examples:** adding ts example and refactoring structure ([5c835be](https://github.com/roggervalf/react-appointment-picker/commit/5c835bef562d547cd9d9479312190a48089436f6))
* **ts:** validation ts usage [#10](https://github.com/roggervalf/react-appointment-picker/issues/10) ([ed75283](https://github.com/roggervalf/react-appointment-picker/commit/ed75283929d761b14304889e501183c9ccf22cd6))


### BREAKING CHANGES

* **ts:** new way using callbacks

## [2.0.1](https://github.com/roggervalf/react-appointment-picker/compare/v2.0.0...v2.0.1) (2020-06-19)


### Bug Fixes

* **types:** allow id to be passed as number ([7ee6315](https://github.com/roggervalf/react-appointment-picker/commit/7ee63151703fa1a4b16f030f8780399e96f6d4a2))

# [2.0.0](https://github.com/roggervalf/react-appointment-picker/compare/v1.2.1...v2.0.0) (2020-06-17)


### Features

* **types:** new refactor adding types [#9](https://github.com/roggervalf/react-appointment-picker/issues/9) ([c000cda](https://github.com/roggervalf/react-appointment-picker/commit/c000cdac60db7bbb8c73c8d694687376cb18c40e))
* **typescript:** adding support for typescript ([8d5ca27](https://github.com/roggervalf/react-appointment-picker/commit/8d5ca27a885f8a4ea54e28d23758a9a6f3694ec7))


### BREAKING CHANGES

* **types:** Support for Typescript

## [1.2.1](https://github.com/roggervalf/react-appointment-picker/compare/v1.2.0...v1.2.1) (2020-04-12)


### Bug Fixes

* **github:** updating github references and adding contributing docs ([e366d96](https://github.com/roggervalf/react-appointment-picker/commit/e366d96f463c39648a4ea96a2ed945a0324c0c2a))

# [1.2.0](https://github.com/roggervalf/react-appointment-picker/compare/v1.1.1...v1.2.0) (2020-03-25)

### Bug Fixes

- **markdownlint:** adding lint for markdown ([e009c5a](https://github.com/roggervalf/react-appointment-picker/commit/e009c5af1c9ac18eba9d3a7c2dd8d5a0e32ace1f))
- **node version:** updating from travis ([55db866](https://github.com/roggervalf/react-appointment-picker/commit/55db866a4fd1cadb2b965d2306854ca5427a0785))
- handlebars vulnerability ([8a7bd8f](https://github.com/roggervalf/react-appointment-picker/commit/8a7bd8fb388f62eb4de74ac182a76c499ceb5175))
- handlebars vulnerability ([caeeaf4](https://github.com/roggervalf/react-appointment-picker/commit/caeeaf4b10bd09295455e91d5d94dfa872e5b4a3))

### Features

- **semantic-release:** automatic deploy and commitizen ([fb33c38](https://github.com/roggervalf/react-appointment-picker/commit/fb33c3873c643b9badd4331624e14d721e023911))

### Performance Improvements

- **commitizen:** adding cm script for applying it ([25293ee](https://github.com/roggervalf/react-appointment-picker/commit/25293ee24364a2d0cb44719b8c6b7c3d9416e2c2))


example-js/src/App.js
import React, { Component } from 'react';

import { AppointmentPicker } from 'react-appointment-picker';

export default class App extends Component {
  state = {
    loading: false,
    continuousLoading: false
  };

  addAppointmentCallback = ({
    addedAppointment: { day, number, time, id },
    addCb
  }) => {
    this.setState(
      {
        loading: true
      },
      async () => {
        await new Promise((resolve) => setTimeout(resolve, 2000));
        console.log(
          `Added appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        addCb(day, number, time, id);
        this.setState({ loading: false });
      }
    );
  };

  removeAppointmentCallback = ({ day, number, time, id }, removeCb) => {
    this.setState(
      {
        loading: true
      },
      async () => {
        await new Promise((resolve) => setTimeout(resolve, 2000));
        console.log(
          `Removed appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        removeCb(day, number);
        this.setState({ loading: false });
      }
    );
  };

  addAppointmentCallbackContinuousCase = ({
    addedAppointment: { day, number, time, id },
    addCb,
    removedAppointment: params,
    removeCb
  }) => {
    this.setState(
      {
        continuousLoading: true
      },
      async () => {
        if (removeCb) {
          await new Promise((resolve) => setTimeout(resolve, 1000));
          console.log(
            `Removed appointment ${params.number}, day ${params.day}, time ${params.time}, id ${params.id}`
          );
          removeCb(params.day, params.number);
        }
        await new Promise((resolve) => setTimeout(resolve, 1000));
        console.log(
          `Added appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        addCb(day, number, time, id);
        this.setState({ continuousLoading: false });
      }
    );
  };

  removeAppointmentCallbackContinuousCase = (
    { day, number, time, id },
    removeCb
  ) => {
    this.setState(
      {
        continuousLoading: true
      },
      async () => {
        await new Promise((resolve) => setTimeout(resolve, 2000));
        console.log(
          `Removed appointment ${number}, day ${day}, time ${time}, id ${id}`
        );
        removeCb(day, number);
        this.setState({ continuousLoading: false });
      }
    );
  };

  render() {
    const days = [
      [
        { id: 1, number: 1, isSelected: true, periods: 2 },
        { id: 2, number: 2 },
        null,
        { id: 3, number: '3', isReserved: true },
        { id: 4, number: '4' },
        null,
        { id: 5, number: 5 },
        { id: 6, number: 6 }
      ],
      [
        { id: 7, number: 1, isReserved: true, periods: 3 },
        { id: 8, number: 2, isReserved: true },
        null,
        { id: 9, number: '3', isReserved: true },
        { id: 10, number: '4' },
        null,
        { id: 11, number: 5 },
        { id: 12, number: 6 }
      ],
      [
        { id: 13, number: 1 },
        { id: 14, number: 2 },
        null,
        { id: 15, number: 3, isReserved: true },
        { id: 16, number: '4' },
        null,
        { id: 17, number: 5 },
        { id: 18, number: 6 }
      ],
      [
        { id: 19, number: 1 },
        { id: 20, number: 2 },
        null,
        { id: 21, number: 3 },
        { id: 22, number: '4' },
        null,
        { id: 23, number: 5 },
        { id: 24, number: 6 }
      ],
      [
        { id: 25, number: 1, isReserved: true },
        { id: 26, number: 2 },
        null,
        { id: 27, number: '3', isReserved: true },
        { id: 28, number: '4' },
        null,
        { id: 29, number: 5 },
        { id: 30, number: 6, isReserved: true }
      ]
    ];
    const { loading, continuousLoading } = this.state;
    return (
      <div>
        <h1>Appointment Picker</h1>
        <AppointmentPicker
          addAppointmentCallback={this.addAppointmentCallback}
          removeAppointmentCallback={this.removeAppointmentCallback}
          initialDay={new Date('2018-05-05')}
          days={days}
          maxReservableAppointments={3}
          alpha
          visible
          selectedByDefault
          loading={loading}
        />
        <h1>Appointment Picker Continuous Case</h1>
        <AppointmentPicker
          addAppointmentCallback={this.addAppointmentCallbackContinuousCase}
          removeAppointmentCallback={
            this.removeAppointmentCallbackContinuousCase
          }
          initialDay={new Date('2018-05-05')}
          days={days}
          maxReservableAppointments={2}
          alpha
          visible
          selectedByDefault
          loading={continuousLoading}
          continuous
        />
      </div>
    );
  }
}
